mono/mini/mini-runtime.c

```diff
@@ -4510,14 +4511,23 @@ mjit_initialize (void)
 }
 
 static gpointer
-compile_method_inner (MonoMethod *method, MonoDomain *target_domain, gint32 opt, MonoError *error)
+mjit_compile_method_inner (MonoMethod *method, MonoDomain *target_domain, gint32 opt, MonoError *error)
 {
-	if (!g_hasenv("MONO_MJIT"))
-		return mono_jit_compile_method_inner (method, target_domain, opt, NULL, error);
-
+	static __thread gboolean is_mjit_compiling = FALSE;
 	MonoObject *compiler, *method_info, *ret;
 	MonoNativeCodeHandle native_code;
 	gpointer params[4];
+	gboolean bootstrapping;
+
+	bootstrapping = is_mjit_compiling;
```

Comment:
I'm confused. is `bootstrapping` ever true?

Reply:
Yes, when you are compiling the `ICompiler.CompileMethod` for example.

Reply:
hmm, it's still not obvious to me what happens here:
* first call of `mjit_compile_method_inner`, let's say for method `Foo ()` -> `is_mjit_compiling = FALSE`
* `bootstrapping = is_mjit_compiling = FALSE`
* since `!bootstrapping`, we set `is_mjit_compiling = TRUE`
* `bootstrapping` is still `FALSE`, so we are using `mono_runtime_invoke_checked ()`
* that means we end up in `mjit_compile_method_inner` again, this time with `bootstrapping = is_mjit_compiling = TRUE`, and therefore return a "fake compiled entry" of `MethodInfo.ctor()`/`ICompiler.CompileMethod()`. By "fake compiled entry" I mean a function pointer that calls into the interpreter again, and executes said method.
* we are back to the first call of `mjit_compile_method_inner` and compile `Foo()` with the managed JIT (the managed JIT runs in the interpreter).
* since `!bootstrapping` is still the case we unsert `is_mjit_compiling` when we are done for `Foo()`.

So either:
1. we anyway run the managed JIT always in the interpreter, then I don't get the point of this change, as it seems to be a complicated flow of things a lot with no change.
2. Or, I'm missing something. Please explain 🙂

Reply:
@lewurm 
* first call of `mjit_compile_method_inner`, let's say for method `Foo ()` -> `is_mjit_compiling = FALSE`
* `bootstrapping = is_mjit_compiling = FALSE`
* since `!bootstrapping`, we set `is_mjit_compiling = TRUE`
* `bootstrapping` is still `FALSE`, so we are using `mono_runtime_invoke_checked ()` **to invoke `MiniCompiler.CompileMethod(method = "Foo")`**
    * if we end up in `mjit_compile_method_inner` again, it means we are compiling `MiniCompiler.CompileMethod ` (with `MiniCompiler` => bootstrapping) -> we invoke with the Interpreter so we don't end up in `mjit_compile_method_inner` again.
* we are back to the first call of `mjit_compile_method_inner` and compile `Foo()` with the managed JIT **which runs as JITted code because we compiled it while running in the Interpreter**.
* since `!bootstrapping` is still the case we unsert `is_mjit_compiling` when we are done for `Foo()`.